### DISCUSSION TITLE
Best way to install into shared file system on z/OS

### DISCUSSION THREAD

**Discussion by MikeFultonDev** [zopencommunity/meta#465] [General]
I have been doing some reading on installation of software on Linux (https://unix.stackexchange.com/questions/215776/recommended-way-to-install-software-to-usr-local-use-sudo-or-chown) and this seems too 'loose' for my liking.

The recommendation is to perform a 'make install' as a user with elevated authority.

I am curious what others think. 

My proposal would be that we provide a 'zopen clone' function, but that it doesn't run the entire 'clone' job under sudo, but instead would sudo to a particular user for just the minimum sequences required.

I am thinking:

- install software into working tree (e.g. `$HOME/usr/local/zopen`) with standard authority
- `sudo mv /usr/local/zopen /usr/local/backup`
- `sudo mv $HOME/usr/local/zopen /usr/local/zopen`
- `sudo chown OWNER:GROUP /usr/local/zopen`

**Comment by v1gnesh** [comment]
`sudo make install` is a nono.
https://www.youtube.com/watch?v=wFlyUzUVFhw

Also, what are your thoughts on the [XDG spec](https://xdgbasedirectoryspecification.com/)?

Category | Environment Variable | Default Value | FHS Approximation
-- | -- | -- | --
Configuration | $XDG_CONFIG_HOME | $HOME/.config | /etc
Data | $XDG_DATA_HOME | $HOME/.local/share | /usr/share
State | $XDG_STATE_HOME | $HOME/.local/state | /var/lib
Cache | $XDG_CACHE_HOME | $HOME/.cache | /var/cache

**Reply by MikeFultonDev** [reply]
Watched the video. Worth the time

**Reply by MikeFultonDev** [reply]
My primary concern with XDG spec is that if software doesn't embrace it fairly globally, you've just created ANOTHER 4 locations in the home directory and not fixed anything. I see the 'who else does this' list - I don't see vim, but I do see git and nano. Does bash support it? If our bash and git support that, I'd definitely turn it on...
We have moved our .config for zopen into our own /etc so I don't believe this is an issue for 'zopen' itself, but curious what you think?

**Reply by v1gnesh** [reply]
I can't say decisively but I think I'm seeing a preference for XDG. 
The usage of `~/.config` is pretty common. in the devblogs, developer tools, etc I've come across in recent times.

Here's a link to a only-slightly-related post on the topic of where to keep what (as the guy thinks it out for an immutable OS - Fedora Silverblue) - https://blog.verbum.org/2020/08/22/immutable-%e2%86%92-reprovisionable-anti-hysteresis/

On bash & XDG, the only good thing I could find on a quick search was https://hiphish.github.io/blog/2020/12/27/making-bash-xdg-compliant/.

**Comment by mfsysprog** [comment]
What would a 'clone' option do? In my opinion a clone option should be used to do a git clone of the <package>port source to a zopen/dev directory so you can locally build the software using zopen build. Or is that what you intent to do here?

**Reply by mfsysprog** [reply]
And zopen install already has an option to set the prefix for installation, so if you set that to / wouldn't the current zopen install already do what you are looking for here?

**Reply by MikeFultonDev** [reply]
ah. perhaps clone is a crappy verb name! The intent is that it's just a 'copy' of a tested install from a sandbox location to the 'official' location on the system to be shared - typically I would expect /usr/local

**Reply by MikeFultonDev** [reply]
@DevonianTeuchter and I have been chatting and we are thinking 'mirror' or (even better) 'promote' might be the right verbs.

**Reply by v1gnesh** [reply]
+1 for promote.

**Reply by DevonianTeuchter** [reply]
"Promote" was what we were thinking also :-)

**Comment by MikeFultonDev** [comment]
Another question - if we do a 'promote' to a shared area, should we strip out the 'zopen' tool since this shouldn't be tampered with in any way, or should we keep it so things like 'list --installed' would function (i'm leaning to 'strip' to prevent having multiple zopen tools lying on the system but curious what folks think

**Reply by lbdyck** [reply]
A promote should be only the production elements and thus exclude `zopen` in my view

**Reply by v1gnesh** [reply]
+1 for leaving it out, however, would adding `list --installed-at <location>`, or a similar generalizable arg help?

**Reply by MikeFultonDev** [reply]
that's an interesting approach @v1gnesh - have zopen package manager be able to 'point' to an alternate location

**Reply by v1gnesh** [reply]
Yeah, it'll be a neat way of managing deployed zopen tooling.

**Reply by lbdyck** [reply]
Would be nice to have a `promote` option and an `upgrade-promote` option where the `upgrade-promote` option would just upgrade the packages already promoted.  I can see some using this.  My plan would be to `promote` to a new zfs and then change the mounted zfs to the new zfs.

**Comment by DevonianTeuchter** [comment]
> Would be nice to have a promote option and an upgrade-promote option where the upgrade-promote option would just upgrade the packages already promoted. I can see some using this. My plan would be to promote to a new zfs and then change the mounted zfs to the new zfs.

The thinking is that if you are promoting to a location (without the zopen admin tooling), you wouldn't really want to make changes to it; even upgrading those packages could, if there had been a build between testing a particular upgrade and the actual upgrade occurring, lead to untested versions in the "promoted" environment. 
Your "_plan would be to promote to a new zfs and then change the mounted zfs to the new zfs._" would really be the usecase for this - a known "good" version promoted to a location [even overwriting/purging an existing location]

**Reply by lbdyck** [reply]
Actually I would expect the use case scenario to be:

1. install into zopen and validate
2. promote into a dev/test zfs for broader testing and validation
3. after (2) then promote into production zfs

**Reply by DevonianTeuchter** [reply]
having something like "`zopen promote --from <testzfs> <promotetarget>`" in other words. Makes sense.

**Reply by MikeFultonDev** [reply]
+1

**Comment by MikeFultonDev** [comment]
@lbdyck @mfsysprog  et al for 'promotion into production zfs' that Lionell mentioned, I was thinking the flow could be something like:

- mount file system to temporary location
- copy files into file system at temporary location
- unmount file system
- unmount /usr/local file system
- mount new file system under /usr/local
- update BPXPRMxx to reflect the new mount location

The question is, should we have a very simple script that runs via sudo for the privileged operations, or to instead have a small number of commands that run under sudo - or perhaps a combination of the two (a small script that sudo's to a small number of commands)?

**Reply by lbdyck** [reply]
most shops will require a change management record for steps 3-5 and you left off step 6 to update the BPXPRMxx in PARMLIB so the mount is now permanent to the new zfs. 

while that can be scripted to some extent I'm not sure it should be.

my $0.01

**Reply by MikeFultonDev** [reply]
ok - so that would push towards using sudo for individual commands (or having someone just su / log in to the system to do it manually.

Added in update to BPXPRMxx 😁

**Reply by MikeFultonDev** [reply]
@lbdyck can you elaborate on the 'change management record' ? I expect this is company specific? What's involved?

**Reply by lbdyck** [reply]
Every shop has change management procedures that require documenting the change, the steps for the change, confirm the changes have been tested in a test/dev environment, and have been signed off by someone as having been validated. Then the change review board/committee reviews that change along with all other requested changes and schedules the change.

Some installations have gone to the extent of having the change implemented by one person, validated by another, and then signed off by another.

**Reply by mfsysprog** [reply]
You're assuming here that zopen will be the only thing under /usr/local. I would just automate steps one and two and leave the rest to the client. Most of the time the zfs is prepared on an installation system and subsequently distributed to a number of lpars in my experience.

**Reply by mfsysprog** [reply]
And that is mostly done with an IPL or at some specific moment and generally a change record will describe the intended changes. This is part of the ITIL processes most companies use.

**Reply by lbdyck** [reply]
And for the BPXPRMxx update I strongly recommend using this command via a z/OS console or via SDSF:

`SETOMVS SYNTAXCHECK=(xx)`

To do a syntax check after updating the member and before setting it permanently. 

Also, some shops will only allow this at an IPL, or during a scheduled maintenance window, so each site needs to be aware of their change management policies.

**Reply by MikeFultonDev** [reply]
@mfsysprog you are right - i was making that assumption. 
But - reading through the steps, does it change them in any way? 
For 'distribution', is this a combination of a parallel sysplex where other systems in the plex would mount the file system read only, and systems that are not part of the plex get it via something else (like sftp) ? 
Is anyone using Ansible for this for the 'distribution' part?

**Reply by MikeFultonDev** [reply]
ok. so in fact the sequence for many shops may be more like:
- mount 'empty' file system to temporary location
- copy files in the file system at temporary location from all tool sources (not just open tools - likely others)
- update BPXPRMxx to reflect the new mount location
- formal review of changes
- activate changes on next IPL

**Reply by lbdyck** [reply]
no - there could be a filesystem that is exclusive to the zopen tools and the activation may occur outside an IPL but during a scheduled maintenance window which may or may not include an IPL

**Reply by MikeFultonDev** [reply]
ok - got it - so a maintenance window may or may not involve an IPL, but you can be confident processes are not running that will be referencing /usr/local so it's safe to 'fiddle with it' correct?

**Reply by lbdyck** [reply]
correct

**Reply by lbdyck** [reply]
and it would be implemented on one, or more systems, during that window.

**Reply by v1gnesh** [reply]
> @lbdyck can you elaborate on the 'change management record' ? I expect this is company specific? What's involved?

@MikeFultonDev How do I sign up to live in this world? :grin: 

It's true that there's change mgmt everywhere, however, providing a bunch of commands or scripts as handy reference doesn't hurt.
IMO, it's a way of demonstrating tools/techniques that can be used.
Pretty sure that many places use the same methods as they devised decades ago; so seeing options is good.

**Comment by lbdyck** [comment]
> You're assuming here that zopen will be the only thing under /usr/local. I would just automate steps one and two and leave the rest to the client. Most of the time the zfs is prepared on an installation system and subsequently distributed to a number of lpars in my experience.

I agree - remember it may **not** be `/usr/local/` but could be something else or could be `/usr/local/tools` as an example.

**Reply by MikeFultonDev** [reply]
it could be some other location than /usr/local, but I would hope most folks could use the default location. I would say one of the biggest issues we have on z/OS is customization (also one of the biggest strengths unfortunately).

**Reply by lbdyck** [reply]
Remember that these tools are being installed on an existing system and they may have tools from multiple sources which they keep separated for various reasons - including being able to update one tool or one set of tools without affecting others.

**Reply by v1gnesh** [reply]
You're very likely to come across companies using `/<sysres>/whatever`, so `/usr/local`, I'm afraid will be an immediate no from the OMVS admins/sysprogs.
